nexus/auth/src/authz/roles.rs

```diff
@@ -82,6 +82,28 @@ impl RoleSet {
             String::from(role_name),
         ));
     }
+
+    /// Effective, i.e., strongest, role in set for specified resource
+    pub fn effective_role(
+        &self,
+        resource_type: &ResourceType,
+        resource_id: &Uuid,
+    ) -> Option<String> {
+        self.roles
+            .iter()
+            .filter(|(typ, id, _)| typ == resource_type && id == resource_id)
+            .max_by_key(|(_, _, role)| role_num(role))
+            .map(|(_, _, role)| role.clone())
+    }
+}
+
+fn role_num(role: &str) -> u8 {
+    match role {
+        "viewer" => 1,
+        "collaborator" => 2,
+        "admin" => 3,
+        _ => 0,
+    }
```

Comment:
this is the worst part

Reply:
My other random PR https://github.com/oxidecomputer/omicron/pull/7477 is relevant here, especially the suggestion of representing roles as an enum rather than a string (now that we know roles don't need to be free-form), which would make this a lot less hacky.

https://github.com/oxidecomputer/omicron/pull/7477#discussion_r1942129596